### ISSUE TITLE
Build fails with VS 2019

### ISSUE THREAD

**Issue by moshg** [fschutt/azul#163]
I cannot build azul.

## Description

When I run `cargo build` on powershell, my shell displays

```
error: failed to run custom build command for `harfbuzz-sys v0.3.0 (https://github.com/maps4print/azul-dependencies?rev=380b7e7cba8b728a3fc89fe28952e3b07aa624e7#380b7e7c)`
process didn't exit successfully: `C:\Users\Shogo\Projects\azul\target\debug\build\harfbuzz-sys-edd981c7a9cf27bd\build-script-build` (exit code: 101)
--- stdout
cargo:rerun-if-env-changed=HARFBUZZ_SYS_NO_PKG_CONFIG

--- stderr
thread 'main' panicked at '

couldn't determine visual studio generator
if VisualStudio is installed, however, consider running the appropriate vcvars script before building this crate
', C:\Users\Shogo\.cargo\git\checkouts\azul-dependencies-70bb1f94316762f9\380b7e7\cmake-0.1.35\src\lib.rs:687:25
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

With VS 2019's `vcvars64`,

```
error: failed to run custom build command for `harfbuzz-sys v0.3.0 (https://github.com/maps4print/azul-dependencies?rev=380b7e7cba8b728a3fc89fe28952e3b07aa624e7#380b7e7c)`
process didn't exit successfully: `C:\Users\Shogo\Projects\azul\target\debug\build\harfbuzz-sys-edd981c7a9cf27bd\build-script-build` (exit code: 101)
--- stdout
cargo:rerun-if-env-changed=HARFBUZZ_SYS_NO_PKG_CONFIG

--- stderr
thread 'main' panicked at '

unsupported or unknown VisualStudio version: 16.0
if another version is installed consider running the appropriate vcvars script before building this crate
', C:\Users\Shogo\.cargo\git\checkouts\azul-dependencies-70bb1f94316762f9\380b7e7\cmake-0.1.35\src\lib.rs:687:25
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

## Version / OS

* azul version:
  since 7add48d77f33e517d11df57d0a76366c3c51aa8e

* Operating system:
  Windows 10 64bit

**Comment by atoav** [comment]
Had precisely the same issue, also Windows 10 x64 and VS 2019 with `vcvars64`

**Comment by sandouli** [comment]
I had a similar issue with Harfbuzz on Windows 10 and resolved it by updating Visual Studio to the latest version and restarting the computer.

**Comment by simonhaines** [comment]
The azul-dependencies repo needs to be updated to include cmake-0.1.39 and cc-1.0.36 to recognise VS2019.

**Comment by yorodm** [comment]
I stated some counterpoints in a related issue, I just want to add them here for your consideration too.

- No need to update cmake, vcpkg already knows how to find all kinds of Visual Studios.

- It is possible to use vcpkg as an alternative in msvc builds without breaking portability.

- There's this vcpkg crate and it seems to be widely used.

**Comment by wycats** [comment]
What needs to happen to get `azul-dependencies` updated?

**Comment by fschutt** [comment]
@wycats someone needs to update it, duh. :wink: Haven't had the time to do that though. Yes, I'll fix it, it's on my list, it's not a huge issue - but I'm currently more concerned with fixing the layout system and fixing segfaults in the OpenGL code. After that's done I'll fix the build issues.

**Comment by darkskygit** [comment]
waiting for your update 😉

**Comment by insomnyawolf** [comment]
Same here, i hope everything is going well.

**Comment by Adarma** [comment]
> it's not a huge issue

I am new to Azul and would like to try it out.
I hope the fix is coming soon, as for me it is a blocking issue.

**Comment by ravaa** [comment]
this is also failing for me

**Comment by fschutt** [comment]
You can try the `/unvendor_dependencies` branch, but that one isn't compiling yet. I'm in the progress of fixing it, don't worry.

**Comment by Mazwak** [comment]
I don't understand why this was closed. It is not fixed yet.

It would have save me 2 hours, as I checked opened issues before trying to fix my installation. I only now checked closed issue to realise that VS2019 is not supported.

**Comment by xKrebs** [comment]
I've this problem..
error: failed to run custom build command for `harfbuzz-sys v0.3.0 (https://github.com/maps4print/azul-dependencies?rev=bf2933b9aac43a7003278862772250398e4fa457#bf2933b9)`

Caused by:
  process didn't exit successfully: `C:\Users\Cristian\Documents\Rust\Pfile\target\debug\build\harfbuzz-sys-2784050289f4621a\build-script-build` (exit code: 101)
--- stdout
cargo:rerun-if-env-changed=HARFBUZZ_SYS_NO_PKG_CONFIG
running: "cmake" "C:\\Users\\Cristian\\.cargo\\git\\checkouts\\azul-dependencies-70bb1f94316762f9\\bf2933b\\harfbuzz-sys-0.3.0\\harfbuzz" "-G" "Visual Studio 16 2019" "-Thost=x64" "-Ax64" "-DCMAKE_INSTALL_PREFIX=C:\\Users\\Cristian\\Documents\\Rust\\Pfile\\target\\debug\\build\\harfbuzz-sys-37d203667a9d0f13\\out" "-DCMAKE_C_FLAGS= -nologo -MD -Brepro" "-DCMAKE_C_FLAGS_DEBUG= -nologo -MD -Brepro" "-DCMAKE_CXX_FLAGS= -nologo -MD -Brepro" "-DCMAKE_CXX_FLAGS_DEBUG= -nologo -MD -Brepro" "-DCMAKE_BUILD_TYPE=Debug"

--- stderr
thread 'main' panicked at '
failed to execute command: Impossibile trovare il file specificato. (os error 2)
is `cmake` not installed?

build script failed, must exit now', C:\Users\Cristian\.cargo\git\checkouts\azul-dependencies-70bb1f94316762f9\bf2933b\cmake-0.1.40\src\lib.rs:835:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

warning: build failed, waiting for other jobs to finish...
error: build failed

**Comment by phanirithvij** [comment]
@fschutt does azul work with `VS2017`?

**Comment by Whimfoome** [comment]
I fixed mine by installing **Visuals Studio Build Tools 2015**, it has to be **2015** and it works.

**Comment by natanfudge** [comment]
>  it's not a huge issue

Am I reading correctly that the library does not build at all is "not a huge issue"? Or perhaps I am misunderstanding

**Comment by nintha** [comment]
There is a temp way to workaround. 
1. download azul project code as zip(default name is `azul-master`), and unzip it under your project root folder.
2. open `azul-master/cargo/azul/Cargo.toml`, change `azul-dependencies       = { version = "0.1.0",                git = "https://github.com/maps4print/azul-dependencies", rev = "..." }` to `azul-dependencies       = { version = "0.1.0",                git = "https://github.com/maps4print/azul-dependencies", rev = "b61fe8d931ea5f46fdce05944f5fde8ca4cd6931" }`.
3.  open `azul-master/cargo/azul-widgets/Cargo.toml`, change `azul-dependencies       = { version = "0.1.0",                git = "https://github.com/maps4print/azul-dependencies", rev = "..." }` to `azul-dependencies       = { version = "0.1.0",                git = "https://github.com/maps4print/azul-dependencies", rev = "b61fe8d931ea5f46fdce05944f5fde8ca4cd6931" }`.
4. edit your project `Cargo.toml`
```toml
[dependencies]
azul = { path = "azul-master/cargo/azul" }
```
5. try to build
```
cargo build
```

**Comment by Mazwak** [comment]
> > it's not a huge issue
> 
> Am I reading correctly that the library not building at all is "not a huge issue"? Or perhaps I am misunderstanding

It does not build on windows with the latest Visual Studio. It does compile on other platforms or on previous versions of Visual Studio.

I do agree it's not a huge issue.

**Comment by GrassLake** [comment]
nintha's reply is awesome!
Just follow his instruction and it works!

**Comment by niklasha** [comment]
Does nintha's method really still work?  I forked azul, changed the 2 dependency lines, and changed my dependent project to use my forked version.  I can see in my Cargo.lock that I have got the b61f.. version but it still fails to build harfbuzz (who needs a text shaper anyway ;-) ).  I have tried VS 2019 16.4.2 and now also 16.4.3.  Of course I can install (and I probably will) install VS2015 instead, but it is intriguing what is causing this.
